kernel/h1/src/pmu.rs

```diff
@@ -269,3 +276,45 @@ pub fn reset_dcrypto() {
     // Clear the DCRYPTO bit, which is 0x2
     unsafe {pmu.reset.set(pmu.reset0.get() & !(0x2));}
 }
+
+pub struct Reset {
+    // The last reset source.
+    reset_source: u8,
+}
+
+impl Reset {
+    const fn new() -> Reset {
+        Reset {
+            reset_source: 0,
+        }
+    }
+
+    pub fn init(&mut self) {
+        let pmu: &mut PMURegisters = unsafe { transmute(PMU) };
+
+        // Read and reset the reset source
+        self.reset_source = unsafe{(pmu.reset_source.get() & 0xff) as u8};
+        unsafe{pmu.clear_reset.set(1)};
+    }
+}
+
+impl reset::Reset for Reset {
+    fn reset(&self) -> ! {
+        let pmu: &mut PMURegisters = unsafe { transmute(PMU) };
+
+        unsafe {pmu.global_reset.set(GLOBAL_RESET_VALUE)};
+
+        // Wait for reboot; should never return
+        loop {
+            unsafe {
+                asm!("dsb" :::: "volatile");
+                asm!("wfi" :::: "volatile");
```

Comment:
The casual reader (me) has no idea at all what these instructions are doing.

Furthermore, a loop that spins, executing two instructions one after another forever seems wrong.  Presumably that's not what it's doing, because presumably the sequence causes the machine to reset, but I had to infer that.  There should at least be a comment that says that.

Finally, if the two-instruction sequence should always cause a reset, then this is not the way I would write it -- I would write it as something like `dsb`; `wfi`; `UNREACHABLE()`.

Either that or I have no clue what's going on here.

Reply:
I actually got that sequence from some of our existing code for this chip. I agree it's a little confusing, but I don't think it's exciting enough. It certainly does what we want it to do ...

Reply:
I guess the comment above the loop gives enough of a high-level view of what's going on, although I'd love to know why it needs to be in a loop.  Perhaps it's written that way to convince the compiler that it never returns?